django/db/models/expressions.py

```diff
@@ -1075,6 +1079,12 @@ def resolve_expression(
             c.source_expressions[pos] = arg.resolve_expression(
                 query, allow_joins, reuse, summarize, for_save
             )
+        if not self.allows_composite_expressions and any(
+            isinstance(expr, ColPairs) for expr in c.get_source_expressions()
+        ):
+            raise ValueError(
+                f"{self.__class__.__name__} does not support composite primary keys."
+            )
```

Comment:
I first tried putting this on `BaseExpression` but then noticed that several methods (including this one) don't call their super implementations. So there might be some more cases to explore. In the meantime, I removed the duplication of this code in BaseExpression given that I didn't have a test case to support it. Happy to hear advice.